### PR TITLE
Power Cell Refactor

### LINKED ISSUE #5751: Laser guns accept any kind of power cell and can become stupid overpowered.
## Description

How to replicate:

- Change the map to Packed Station.
- Go to Cargo and order a Laser Crate.
- Open the crate and get a Retro Laser Gun.
- Go to Science and get the Large Power Cell they have there.
- Your Retro Laser now has 225 shots in it. 
- This also applies (to a lesser degree) to the lasers in the armory on Packed.

Logically, each gun should only take one cell type (small or medium, as large is supposed to be for APCs as I understand it) and be unable to take anything else.

### PR DESCRIPTION
**Pull request by ElectroJr**
## About the PR

This is a refactor of power cells, cell-slots, and chargers with the goal to:
- Remove component inheritance
- Standardise power cell eject/insert interactions (less code duplication)
- ECS power cells

Some files were moved to shared, and others were just renamed (e.g. `BaseCharger.cs` became `ChargerComponent.cs`. Despite using `git mv` in a separate commit, github diff seems to show these as different files, which is inflating the size of the PR a bit.

Changes to chargers:
- Previously `WeaponCapacitorCharger` and `PowerCellCharger` inherited from `BaseCharger`
- They are now a single `Charger` component + system (not fully ECS)
- Chargers now use `ItemSlots` for insert/eject interactions.

Changes to power cells:
- `PowerCellComponent` no longer inherits from `BatteryComponent`.
- `PowerCellSlotComponent` now use `ItemSlots` for insert/eject interactions.
- Logic in cell & cell-slot components all moved to `PowerCellSystem`
- `ServerBatteryBarrelComponent` now uses the `PowerCellSlot` component instead of duplicating all the cell-insert logic.

fixes #5751

:cl:
- fix: Energy weapons now only accept small power cells.

### CODE REVIEW

File: Content.Shared/PowerCell/Components/PowerCellSlotComponent.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    public bool StartEmpty = false;
+
+    /// <summary>
+    /// String passed to <see><cref>String.Format</cref></see> when showing the description text for this item.
+    /// String.Format is given a single parameter which is the size letter (S/M/L) of the cells this component uses.
+    /// Use null to show no text.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("descFormatString")]
+    public string? DescFormatString { get; set; } = "It uses size {0} power cells.";
```
Comment (mirrorcult) [inline Content.Shared/PowerCell/Components/PowerCellSlotComponent.cs:58]: localize

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Battery/battery_guns.yml [hunk 2]
```diff
@@ -16,16 +16,11 @@
     - Single
     soundGunshot:
       path: /Audio/Weapons/Guns/Gunshots/laser.ogg
+  - type: PowerCellSlot
     cellSlot:
       ejectOnUse: true
-      insertSound: /Audio/Weapons/Guns/MagIn/revolver_magin.ogg
-      ejectSound: /Audio/Weapons/Guns/MagOut/revolver_magout.ogg
       soundOptions: 
         volume: -2
```
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Battery/battery_guns.yml:23]: why remove the insert/eject sounds?

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Battery/battery_guns.yml (hunk 2 above)
Comment (ShadowCommander) [reply Resources/Prototypes/Entities/Objects/Weapons/Guns/Battery/battery_guns.yml:23]: They got set as defaults in the C# file.

**Review by ShadowCommander — CHANGES_REQUESTED** [review]

**Review by ShadowCommander — APPROVED** [review]

**Comment by metalgearsloth** [comment]
Conflicts and merge

**Pull request merged**